Rocket.Chat/Models/ChatNotification.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  ChatNotification.swift
+//  Rocket.Chat
+//
+//  Created by Samar Sunkaria on 4/8/18.
+//  Copyright © 2018 Rocket.Chat. All rights reserved.
+//
+
+import Foundation
+
+struct ChatNotification: Equatable {
+    var sender: Sender
+    var type: NotificationType
+    var title: String
+    var body: String
+    var rid: String
+}
+
+extension ChatNotification {
+    // swiftlint:disable nesting
+    struct Sender: Codable, Equatable {
+        let name: String
+        let username: String
+        let id: String
+
+        enum CodingKeys: String, CodingKey {
+            case name
+            case username
+            case id = "_id"
+        }
+    }
+
+    typealias ChannelName = String
+
+    enum NotificationType: Equatable {
+        case channel(ChannelName)
+        case group(ChannelName)
+        case direct(Sender)
+    }
+}
+
+extension ChatNotification: Decodable {
+    enum CodingKeys: String, CodingKey {
+        case title
+        case body = "text"
+        case payload
+    }
+
+    enum PayloadKeys: String, CodingKey {
+        case sender
+        case channelName = "name"
+        case type
+        case rid
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.title = try container.decode(String.self, forKey: .title)
+        self.body = try container.decode(String.self, forKey: .body).trimmingCharacters(in: .newlines)
```

Comment:
Question: Can you avoid this manual decoding by following the JSON structure and having a Payload struct as a property of ChatNotification?

Manual decoding is tricky to maintain, so if you could avoid it as much as possible, it would be good :+1:

Reply:
The only reason I decided to go for manual decoding is to allow for this enum. 

```
enum NotificationType: Equatable {
        case channel(ChannelName)
        case group(ChannelName)
        case direct(Sender)
 }
```

Which makes allows me to have a rigid structure with no optionals. 

I can change it to match the structure of the notification, but that will introduce some if-else logic and optional checking elsewhere in the code.

Reply:
@cardoso I think I've found a middle ground here. Now, `payload` is a part of the `ChatNotification` struct and only it gets a custom decoder. Check the latest commit.